packages/core/src/EntityManager.ts

```diff
@@ -475,8 +475,11 @@ export class EntityManager<D extends IDatabaseDriver = IDatabaseDriver> {
       return entity;
     }
 
+    const meta = this.metadata.get(entityName);
+    const childMeta = this.metadata.getByDiscriminatorColumn(meta, data as EntityData<T>);
+
     entity = Utils.isEntity<T>(data) ? data : this.getEntityFactory().create<T>(entityName, data as EntityData<T>, { merge: true, refresh, convertCustomTypes });
-    this.validator.validate(entity, data, this.metadata.find(entityName)!);
+    this.validator.validate(entity, data, childMeta !== undefined ? childMeta : meta);
```

Comment:
the original approach with `??` was better :] 

```suggestion
    this.validator.validate(entity, data, childMeta ?? meta);
```